src/loader/bgw_launcher.c

```diff
@@ -282,23 +293,50 @@ populate_database_htab(void)
 								 * connections or are templates */
 
 		db_oid = HeapTupleGetOid(tup);
-		db_he = (DbHashEntry *) hash_search(db_htab, &db_oid, HASH_ENTER, &hash_found);
+		(void) hash_search(staging_db_htab, &db_oid, HASH_ENTER, &hash_found);
+	}
+	heap_endscan(scan);
+
+	/*
+	 * Now reserve slots for all schedulers via the bgw_counter. We want to
+	 * avoid the race condition where we have enough workers allocated to
+	 * start schedulers for all databases, but before we could get all of them
+	 * started, the (say) first scheduler has started too many jobs and then
+	 * we don't have enough schedulers for the dbs. So we need to increment
+	 * our workers all at once so that schedulers don't start workers stealing
+	 * other schedulers' spots.
+	 */
+	if (!bgw_total_workers_increment_by(hash_get_num_entries(staging_db_htab)))
+	{
+		ereport(LOG, (errcode(ERRCODE_CONFIGURATION_LIMIT_EXCEEDED),
+					  errmsg("total databases = %ld TimescaleDB background worker limit %d, so no schedulers allocated", hash_get_num_entries(db_htab), guc_max_background_workers),
+					  errhint("You may start background workers manually by using the  _timescaledb_internal.start_background_workers() function in each database you would like to have a scheduler worker in.")));
+		return db_htab;
+	}
+
+	/*
+	 * All schedulers have been correctly accounted for, so go ahead and
+	 * insert all into the actual hash table.
+	 */
+	hash_seq_init(&hash_seq, staging_db_htab);
+
+	while ((current_entry = hash_seq_search(&hash_seq)) != NULL)
+	{
+		DbHashEntry *db_he;
+		bool		hash_found;
+
+		db_he = (DbHashEntry *) hash_search(db_htab, &current_entry->db_oid, HASH_ENTER, &hash_found);
 		if (!hash_found)
 			db_he->db_scheduler_handle = NULL;
 	}
-	heap_endscan(scan);
+	hash_destroy(staging_db_htab);
+
 	heap_close(rel, AccessShareLock);
 	CommitTransactionCommand();
 	return db_htab;
 }
 
-/*
- * We want to avoid the race condition where we have enough workers allocated to start schedulers
- * for all databases, but before we could get all of them started, the (say) first scheduler has
- * started too many jobs and then we don't have enough schedulers for the dbs. So we need to first
- * get the number of dbs for which we may need schedulers, then reserve the correct number of workers,
- * then start the workers.
- */
+/* Scan our hash table of dbs and register a worker for each */
```

Comment:
This function should only be called immediately after `populate_db_htab`, should it, and the call to `before_shmem_exit` be called directly from `populate_db_htab`?

Reply:
I don't think it matters? `before_shmem_exit` just registers the hook, which is fine right after + outside of `populate_db_htab`. I think it's fine if `start_db_schedulers` is called after `before_shmem_exit` (in fact I think `before_shmem_exit` should be ordered before `start_db_schedulers`, but I think you agree). What is your concern here?

Reply:
Just code clarity; it seems like `populate_db_htab` and `start_db_schedulers` are rather tightly coupled.

Reply:
I think it's ok. They're not necessarily tightly coupled? All the accounting checks out even if there was some extended period of time between `populate_db_htab` and `start_db_schedulers`